In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statistics as stats
import scipy as sts

In [3]:
# T-shirts sales per day by X and Y company . 

X=[50,123,87,150,68,49,75]
Y=[120,160,90,145,105,85,100]

# Coefficient of Variation

In [17]:
# **finding consistence**
# measure of central tendency and measure of dispersion

# X company 
mean_x=np.mean(X)
var_x=np.var(X)

sd_x=np.std(X) # or var_x**0.5

# Y company
mean_y=np.mean(Y)
var_y=np.var(Y)

sd_y=np.std(Y) # or var_x**0.5

# Coefficient of Variation for both the Company
X_CV=sd_x/mean_x

Y_CV=sd_y/mean_y

print (f'Coefficient of Variation for X is : {X_CV}') 
print('\n')
print (f'Coefficient of Variation for X is : {Y_CV}')

print('\n')
if X_CV > Y_CV :
    print('Y Company had more Consistence with less variation and Stability Distribution ')
else:
    print('X Company had more Consistence with less variation and Stability Distribution ')


Coefficient of Variation for X is : 0.4070004739166393


Coefficient of Variation for X is : 0.227705624594079


Y Company had more Consistence with less variation and Stability Distribution 


# Quantile, Percentile

In [24]:
# the math exam scores of 20 students:
#Find the score above which 65% of students scored.
scores = [45, 60, 72, 58, 81, 77, 69, 88, 54, 63, 
          90, 85, 70, 52, 66, 74, 95, 80, 61, 68]

#That means we want the 35th percentile of the scores.
scores.sort()

# compute 35th percentile using 'quantile'
# As 35th percentile divides the data in 35-65%
Student_scores = pd.Series(scores)
scr=Student_scores.quantile(0.35)

print(f'The score above which 65% of students scored is about {scr}')

The score above which 65% of students scored is about 64.95


In [69]:
# Percentile
print(f'The score above which 65% of students scored is about {np.percentile(scores,35)}')

The score above which 65% of students scored is about 64.95


In [44]:
# OR (n+1)*p where n is len of values (or) no. of scores 
scores

[45,
 52,
 54,
 58,
 60,
 61,
 63,
 66,
 68,
 69,
 70,
 72,
 74,
 77,
 80,
 81,
 85,
 88,
 90,
 95]

In [45]:
scores[6:8]

[63, 66]

In [38]:
# (n+1)*p
.35*(20+1)

7.35

In [46]:
63+(.35*(66-63))
# both way can do 

64.05

In [25]:
# consider the series of wages
wage = pd.Series([247, 228, 352, 298, 345, 312, 300, 285, 246, 352, 297, 304, 265, 325, 276])

# compute 35th percentile using 'quantile'
# As 35th percentile divides the data in 35-65%
wage.quantile(0.35)

284.1

In [49]:
wage= wage.sort_values(ascending=True).reset_index(drop=True)

In [55]:
len(wage)

15

In [60]:
# 35th percentile or .35 quantile
# (n+1)*p in some textbook and default use in numpy,pandas,sklearn (n-1)*p
n=len(wage)
p=.35
index_val=(n-1)*p

print(index_val)


4.8999999999999995


In [59]:
wage[4:6]

4    276
5    285
dtype: int64

In [67]:
# 5th index value + .6 * (5th index val - 6th index val)

above_35 = wage[4] + 0.9*(wage[5] - wage[4]) 

print(above_35)

284.1


# Real world example

In [74]:
# last month’s customer spending data (in dollars)

spend = pd.Series([80, 95, 120, 130, 150, 200, 250, 300, 110, 170, 90, 220, 400, 130, 140])


threshold = spend.quantile(0.40)
print("30th Percentile:", threshold)

print(f'That means 40% of customers spent ≤ ${threshold}')

30th Percentile: 130.0
That means 40% of customers spent ≤ $130.0


In [75]:
high_value = spend[spend > threshold]
low_value  = spend[spend <= threshold]

print("High-value customers:\n", high_value)
print("\nLow-value customers:\n", low_value)


High-value customers:
 4     150
5     200
6     250
7     300
9     170
11    220
12    400
14    140
dtype: int64

Low-value customers:
 0      80
1      95
2     120
3     130
8     110
10     90
13    130
dtype: int64


# Business Impact

High-value customers (60%) → Target with loyalty programs, cashback, VIP offers.

Low-value customers (40%) → Target with discounts & promotions to push them upward.

In [76]:
loans = pd.Series([2.5, 3.0, 4.0, 5.0, 5.5, 6.0, 7.5, 8.0, 9.0, 10.0, 12.0, 15.0, 18.0, 20.0, 25.0])

# Find 35th Percentile

threshold = loans.quantile(0.35)
print("35th Percentile (threshold):", threshold)



35th Percentile (threshold): 5.949999999999999


In [77]:
high_value_loans = loans[loans > threshold]
low_value_loans  = loans[loans <= threshold]

print("High-value loans:\n", high_value_loans)
print("\nLow-value loans:\n", low_value_loans)


High-value loans:
 5      6.0
6      7.5
7      8.0
8      9.0
9     10.0
10    12.0
11    15.0
12    18.0
13    20.0
14    25.0
dtype: float64

Low-value loans:
 0    2.5
1    3.0
2    4.0
3    5.0
4    5.5
dtype: float64


#  Business Impact

High-value loans (65%) → higher returns but higher credit risk.

    - Banks may apply stricter fraud detection & risk scoring.

    - Closer monitoring for early defaults.

Low-value loans (35%) → lower risk but smaller profit.

    - Often used for portfolio balancing.

    - May push cross-selling (credit cards, insurance) to improve profitability.

# One Sample Z test

In [78]:
''' A shoe company claims that the average foot length of adult males is 10.5 inches.
As a data analyst, you collect a large random sample (n > 30) of customer foot lengths to check if the claim is true.'''

import numpy as np
from statsmodels.stats.weightstats import ztest

# Suppose we collected foot length data (inches) from 50 customers
sample = np.array([10.7, 10.4, 10.6, 10.8, 10.3, 10.9, 10.5, 10.6, 10.4, 10.8,
                   10.7, 10.9, 10.6, 10.5, 10.4, 10.7, 10.8, 10.6, 10.5, 10.9,
                   10.4, 10.5, 10.7, 10.8, 10.6, 10.9, 10.5, 10.6, 10.4, 10.8,
                   10.7, 10.6, 10.9, 10.5, 10.4, 10.7, 10.8, 10.6, 10.5, 10.9,
                   10.4, 10.5, 10.7, 10.8, 10.6, 10.9, 10.5, 10.6, 10.4, 10.8])

# Hypothesized population mean (company claim)
mu = 10.5

# Perform One-sample Z-test
z_stat, p_val = ztest(sample, value=mu)

print("Z-statistic:", z_stat)
print("p-value:", p_val)


Z-statistic: 5.39107832350775
p-value: 7.003613846229279e-08


Z-statistic = 5.39

This means your sample mean is 5.39 standard errors away from the claimed population mean.

A value this far away is extremely unlikely if the null hypothesis (H₀) were true.

p-value = 7.0 × 10⁻⁸ (0.00000007)

This is much smaller than 0.05 (5% level) or even 0.01 (1% level).

So you reject the null hypothesis H₀.

# Conclusion

There is very strong statistical evidence that your sample mean is significantly different from the hypothesized population mean.

**In business/data analytics terms** → the company’s claim (e.g., “average height = 170 cm”) is **not supported** by your sample data

# Two Sample Z test


An e-commerce company wants to check if average customer spending differs between Region A and Region B.

Region A sample (n = 60 customers): mean = ₹5200, σ = ₹300

Region B sample (n = 50 customers): mean = ₹5000, σ = ₹280

In [80]:
import math

# Given data
x1, sigma1, n1 = 5200, 300, 60   # Region A
x2, sigma2, n2 = 5000, 280, 50   # Region B

# Z-statistic
z = (x1 - x2) / math.sqrt((sigma1**2 / n1) + (sigma2**2 / n2))

# Two-tailed p-value
from scipy.stats import norm
p = 2 * (1 - norm.cdf(abs(z)))

print("Z-statistic:", z)
print("p-value:", p)


Z-statistic: 3.6107907233705943
p-value: 0.000305264909820524


**Interpretation**

Z = 3.61 → the difference between Region A and Region B means is 3.55 standard errors away.

p-value = 0.000305 < 0.05 → Reject H₀.

**Conclusion**: Average spending in Region A (₹5200) is significantly higher than in Region B (₹5000).

Customers in Region A spend more than those in Region B.

The company could allocate more marketing budget to Region A (higher-value customers) 
or design targeted promotions in Region B to boost spending.

# Paired t-test for Solar Panel Output

In [83]:
import numpy as np
import pandas as pd
from scipy import stats

# Simulated daily solar power output (kWh) for 10 panels
# Before cleaning
before_cleaning = np.array([4.8, 5.0, 4.7, 5.1, 4.9, 4.6, 5.0, 4.8, 4.7, 4.9])

# After cleaning (expected improvement due to reduced dust)
after_cleaning = np.array([5.3, 5.5, 5.1, 5.6, 5.4, 5.0, 5.3, 5.2, 5.1, 5.5])

# Create DataFrame for clarity
df = pd.DataFrame({
    "Before_Cleaning": before_cleaning,
    "After_Cleaning": after_cleaning
})

print("🔹 Sample Data:")
print(df)

# Paired t-test
t_stat, p_val = stats.ttest_rel(after_cleaning, before_cleaning)

print("\n🔹 Results:")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_val:.4f}")

# Interpretation
alpha = 0.05
if p_val < alpha:
    print("✅ Reject Null Hypothesis: Cleaning significantly improved power output.")
else:
    print("❌ Fail to Reject Null Hypothesis: No significant improvement found.")


🔹 Sample Data:
   Before_Cleaning  After_Cleaning
0              4.8             5.3
1              5.0             5.5
2              4.7             5.1
3              5.1             5.6
4              4.9             5.4
5              4.6             5.0
6              5.0             5.3
7              4.8             5.2
8              4.7             5.1
9              4.9             5.5

🔹 Results:
T-statistic: 16.7447
P-value: 0.0000
✅ Reject Null Hypothesis: Cleaning significantly improved power output.


# One Sample T-test

You want to check if the average daily energy output of a solar panel batch is meeting the expected standard of 5 kWh/day (benchmark).

**Null hypothesis (H₀):** The mean daily output = 5 kWh.

**Alternative hypothesis (H₁):** The mean daily output ≠ 5 kWh.

In [85]:
import numpy as np
from scipy import stats

# Daily output data (kWh) from 12 solar panels
daily_output = np.array([4.9, 5.1, 4.8, 5.0, 5.2, 4.7, 5.1, 4.9, 5.3, 5.0, 4.8, 5.2])

# Population mean (benchmark)
mu = 5.0

# One-sample t-test
t_stat, p_val = stats.ttest_1samp(daily_output, mu)

print("🔹 Results:")
print(f"Sample Mean: {daily_output.mean():.3f}")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_val:.4f}")

# Interpretation
alpha = 0.05
if p_val < alpha:
    print("✅ Reject Null Hypothesis: Mean output is significantly different from 5 kWh.")
else:
    print("❌ Fail to Reject Null Hypothesis: Mean output is not significantly different from 5 kWh.")


🔹 Results:
Sample Mean: 5.000
T-statistic: 0.0000
P-value: 1.0000
❌ Fail to Reject Null Hypothesis: Mean output is not significantly different from 5 kWh.


**Interpretation:**

In this case, the mean output (≈5.0 kWh) is very close to the benchmark.

Since p > 0.05, we fail to reject H₀ → the plant is performing as expected.

# Two Sample Independent T-tests

**Compare Two Types of Solar Panels**

Suppose you are testing whether Panel Model A and Panel Model B have different average daily outputs (kWh).

Null hypothesis (H₀): Mean output of Model A = Mean output of Model B.

Alternative hypothesis (H₁): Mean output of Model A ≠ Mean output of Model B.

In [87]:
import numpy as np
from scipy import stats
import pandas as pd

# Daily power output (kWh) from two panel models
model_A = np.array([5.1, 5.0, 4.9, 5.2, 5.3, 5.1, 5.0, 5.2, 4.8, 5.1])
model_B = np.array([4.7, 4.8, 4.9, 4.6, 4.8, 4.7, 4.9, 4.6, 4.8, 4.7])

# Create DataFrame for clarity
df = pd.DataFrame({"Model_A": model_A, "Model_B": model_B})
print("🔹 Sample Data:")
print(df)

# Two-sample independent t-test (assume equal variances first)
t_stat, p_val = stats.ttest_ind(model_A, model_B)

print("\n🔹 Results:")
print(f"Mean Model A: {model_A.mean():.3f}")
print(f"Mean Model B: {model_B.mean():.3f}")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_val:.4f}")

# Interpretation
alpha = 0.05
if p_val < alpha:
    print("✅ Reject Null Hypothesis: Significant difference between Model A and Model B outputs.")
else:
    print("❌ Fail to Reject Null Hypothesis: No significant difference between Model A and Model B.")


🔹 Sample Data:
   Model_A  Model_B
0      5.1      4.7
1      5.0      4.8
2      4.9      4.9
3      5.2      4.6
4      5.3      4.8
5      5.1      4.7
6      5.0      4.9
7      5.2      4.6
8      4.8      4.8
9      5.1      4.7

🔹 Results:
Mean Model A: 5.070
Mean Model B: 4.750
T-statistic: 5.4880
P-value: 0.0000
✅ Reject Null Hypothesis: Significant difference between Model A and Model B outputs.


**Interpretation**

Model A: ~5.07 kWh/day

Model B: ~4.77 kWh/day

p-value < 0.05 → significant difference in average output.

 Conclusion: Model A performs better than Model B.